### PR TITLE
Update compaction planning

### PR DESCRIPTION
**Pull request by jwilder**
* Fixes FileStats not actually getting loaded for MMAP'd files
* Changes compaction planning and naming scheme to updated design
* Avoid scanning index too frequently during TSM compactions

### CODE REVIEW

File: tsdb/engine/tsm1/compact.go [hunk 1]
```diff
@@ -38,63 +45,178 @@ var (
...
-// 1) If there are more than 10 closed WAL segments, it will use the 10 oldest
-// 2) If there are any TSM files that contain points that would be overwritten
-//    by a WAL segment, those TSM files are included
-// 3) If there are fewer than 10 WAL segments and no TSM files are required to be
-//    re-written, any TSM files containing tombstones are included.
-// 4) If thare are still fewer than 10 WAL segments and no TSM files included, any
-//    TSM files less than the max file size are included.
+// DefaultPlanner implements CompactionPlanner using a strategy to roll up
+// multiple generations of TSM files into larger files in stages.  It attempts
+// to minimize the number of TSM files on disk while rolling up a bounder number
```
Comment (otoolep) [inline tsdb/engine/tsm1/compact.go]: bounder -> bounded, I presume.

File: tsdb/engine/tsm1/compact.go [hunk 2]
```diff
@@ -38,63 +45,178 @@ var (
...
+	var max time.Time
+	for _, f := range t.files {
+		if f.LastModified.After(max) {
+			max = f.LastModified
+		}
+	}
+	return max
+}
+
+// count return then number of files in the generation
```
Comment (otoolep) [inline tsdb/engine/tsm1/compact.go]: Nit: return then -> returns the

**Comment by otoolep** [comment]
Generally speaking, I think this change needs some logging added. I think it's important for me, as an operator, to be able to trace the life-cycle of files as they are compacted. What is selected for compaction, what it is compacted to, etc etc.

If you want to wrap it in a "compaction-trace" flag, that's fine.

**Comment by otoolep** [comment]
FYI: I am reviewing individual commits, so it is more tractable for me. This means that some review comments may not be valid, when the PR is viewed as a whole.

**Comment by pauldix** [comment]
@otoolep please don't comment on individual commits, it's very confusing to try to look at. Better to comment on the PR as a whole

File: tsdb/engine/tsm1/compact.go [hunk 3]
```diff
@@ -21,15 +21,22 @@ import (
...
+// exist before a compaction cycle will run
+const minCompactionFileCount = 5
+
+// maxCompactionFileCount is the maximum number of TSM files that can be
+// compacted at one time.  A lower value would shorten
 // compaction times and memory requirements, but produce more TSM files
 // with lower compression ratios.  A higher value increases compaction times
-// and memory usage but produces more dense TSM files.
-const maxCompactionSegments = 10
+// and memory usage but produces more dense TSM files. This value is a cutoff
```
Comment (pauldix) [inline tsdb/engine/tsm1/compact.go]: How is the memory usage greater with more files? I would expect the compaction to run as a merge and at any given time you'd never have more than a few Values blocks in memory.

Comment (jwilder) [inline tsdb/engine/tsm1/compact.go]: You're right.  Comment is incorrect because it was rolled over from when these were WAL segments.

**Comment by jwilder** [comment]
@otoolep Agree w/ @pauldix Commenting on commits is really hard to follow.

**Comment by otoolep** [comment]
OK, I'll keep the comments on the main PR.

**Comment by pauldix** [comment]
Going to merge this for now so @otoolep and I can build on it. We'll address things in the upcoming PRs.
